src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,505 @@
+		iter, err := blk.SeriesIter()
+		if err != nil {
+			return err
+		}
+		depIters[i] = iter
+	}
+
+	bounds := seriesIter.Meta().Bounds
+	steps := int((aggDuration + bounds.Duration) / bounds.StepSize)
+	values := make([]float64, 0, steps)
+
+	seriesMeta := seriesIter.SeriesMeta()
+	resultSeriesMeta := make([]block.SeriesMeta, len(seriesMeta))
+	for i, m := range seriesMeta {
+		tags := m.Tags.WithoutName()
+		resultSeriesMeta[i].Tags = tags
+		resultSeriesMeta[i].Name = tags.ID()
+	}
+
+	builder, err := c.controller.BlockBuilder(seriesIter.Meta(), resultSeriesMeta)
+	if err != nil {
+		return err
+	}
+
+	if err := builder.AddCols(bounds.Steps()); err != nil {
+		return err
+	}
+
+	for seriesIter.Next() {
+		values = values[0:0]
+		for i, iter := range depIters {
+			if !iter.Next() {
+				return fmt.Errorf("incorrect number of series for block: %d", i)
+			}
+
+			s, err := iter.Current()
+			if err != nil {
+				return err
+			}
+
+			values = append(values, s.Values()...)
+		}
+
+		desiredLength := int(aggDuration / bounds.StepSize)
+		series, err := seriesIter.Current()
+		if err != nil {
+			return err
+		}
+
+		for i := 0; i < series.Len(); i++ {
+			val := series.ValueAtStep(i)
+			values = append(values, val)
+			newVal := math.NaN()
+			if desiredLength <= len(values) {
+				values = values[len(values)-desiredLength:]
+				newVal = c.processor.Process(values)
+			}
+
+			builder.AppendValue(i, newVal)
+
+		}
+	}
+
+	nextBlock := builder.Build()
+	defer nextBlock.Close()
+	return c.controller.Process(nextBlock)
+}
+
+func (c *baseNode) sweep(processedKeys []bool, queryStartBounds block.Bounds, queryEndBounds block.Bounds, maxBlocks int) {
+	prevProcessed := 0
+	numBlocks := len(processedKeys) - 1
+	for i := numBlocks; i >= 0; i-- {
+		processed := processedKeys[i]
+		if !processed {
+			prevProcessed = 0
+			continue
+		}
+
+		dependantBlocks := maxBlocks
+		if dependantBlocks > numBlocks-i-1 {
+			dependantBlocks = numBlocks - i - 1
+		}
+
+		if prevProcessed >= dependantBlocks {
+			c.cache.remove(i)
+		}
+
+		prevProcessed++
+
+	}
+}
+
+// Processor is implemented by the underlying transforms
+type Processor interface {
+	Process(values []float64) float64
+}
+
+// MakeProcessor is a way to create a transform
+type MakeProcessor func(op baseOp, controller *transform.Controller) Processor
+
+type processRequest struct {
+	blk    block.Block
+	bounds block.Bounds
+	deps   []block.Block
+}
+
+// blockCache keeps track of blocks from the same parent across time
+type blockCache struct {
+	mu              sync.Mutex
+	initialized     bool
+	blockList       []block.Block
+	op              baseOp
+	transformOpts   transform.Options
+	startBounds     block.Bounds
+	endBounds       block.Bounds
+	processedBlocks []bool
+}
+
+func newBlockCache(op baseOp, transformOpts transform.Options) *blockCache {
+	return &blockCache{
+		op:            op,
+		transformOpts: transformOpts,
+	}
+}
+func (c *blockCache) init(bounds block.Bounds) {
```

Comment:
nit newline

Reply:
if it fails --race test then I will move within mutex